dsgrid/dataset/dataset_schema_handler_standard.py

```diff
@@ -40,10 +51,123 @@ def get_unique_dimension_rows(self):
+        # TODO: prefilter time
+        if context.model.project.drop_dimensions:
+            columns = [x.value for x in context.model.project.drop_dimensions]
+            lk_df = lk_df.drop(*columns)
+
+        for dim_type, project_record_ids in context.iter_record_ids_by_dimension_type():
+            dataset_mapping = self._get_dataset_to_project_mapping_records(dim_type)
+            if dataset_mapping is None:
+                dataset_record_ids = project_record_ids
+            else:
+                dataset_record_ids = (
+                    dataset_mapping.withColumnRenamed("from_id", "dataset_record_id")
+                    .join(
+                        project_record_ids,
+                        on=dataset_mapping.to_id == project_record_ids.id,
+                    )
+                    .select("dataset_record_id")
+                    .distinct()
+                    .withColumnRenamed("dataset_record_id", "id")
+                )
+            if dim_type == self.get_pivot_dimension_type():
+                # Drop columns that don't match requested project record IDs.
+                cols_to_keep = {x.id for x in dataset_record_ids.collect()}
+                cols_to_drop = set(self.get_pivot_dimension_columns()).difference(cols_to_keep)
+                if cols_to_drop:
+                    ld_df = ld_df.drop(*cols_to_drop)
+            else:
+                # Drop rows that don't match requested project record IDs.
+                tmp = dataset_record_ids.withColumnRenamed("id", "dataset_record_id")
+                lk_df = lk_df.join(
+                    tmp,
+                    on=lk_df[dim_type.value] == tmp.dataset_record_id,
+                ).drop("dataset_record_id")
+
+        return lk_df, ld_df
+
+    def _process_metric_aggregations(self, context: QueryContext, ld_df):
+        orig_project_pivot_columns = self.get_pivot_dimension_columns_mapped_to_project()
+        pivot_columns = set(ld_df.columns).intersection(orig_project_pivot_columns)
+        assert (
+            len(context.model.project.metric_reductions) <= 1
+        ), "More than one metric aggregation is not supported yet"
+        # TODO: need to merge dataframes on each loop
+        for aggregation in context.model.project.metric_reductions:
+            query_name = aggregation.query_name
+            dim_type, records = context.get_metric_reduction_records(query_name)
+            column = dim_type.value
+            assert column == self.get_pivot_dimension_type().value, self.get_pivot_dimension_type()
+            ld_df, columns = map_and_reduce_pivot_dimension(
+                ld_df,
+                records,
+                pivot_columns,
+                aggregation.operation,
+                rename=True,
+            )
+            if context.metric_columns is None:
+                context.metric_columns = columns
+            elif context.metric_columns != columns:
```

Comment:
I don't fully follow the purpose of this check.

Reply:
We need to store the value of the metric columns. They could be changed during mapping/aggregation/reduction. Every dataset should produce the same metric columns, so assignment only has to happen once. Then every other iteration should be checked to ensure they match.

A case I'm not sure about is one where the user asks for all end uses. Clearly, resstock and tempo will be missing each other's end uses. What happens to the not-applicable columns? Do we fill with null? We've said "no null values for load", but that was really only for individual datasets. Seems like the final dataset can have them.

Reply:
If a query asks for both ResStock and Tempo, then yes, we will have to fill some of the columns with null, it's an artifact of having pivoted column. In the case of ResStock only, I think we want to show the null columns. I remember in the very beginning of dsgrid that we wanted this type of behaviors so users know they are indeed zeros.

Reply:
Should the value be 0 instead of null? I think it's fair to say that the TEMPO dataset has 0 load for electricity_cooling. The columns could be added at various places. I need to think about it.

Reply:
Will insert nulls for now. Make a note that we have to do something with nulls when the user asks for mean.